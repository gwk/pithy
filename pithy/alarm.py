# Dedicated to the public domain under CC0: https://creativecommons.org/publicdomain/zero/1.0/.

import signal
from typing import Callable, ContextManager

from .typing import OptBaseExc, OptTraceback, OptTypeBaseExc


class Timeout(Exception):
  'Exception indicating that an Alarm timed out.'


class Alarm(ContextManager):

  def __init__(self, timeout:int, msg:str='alarm timed out after {timeout} seconds', on_signal:Callable[[],None]|None=None):
    if not isinstance(timeout, int): raise TypeError(f'timeout must be an int; received: {timeout!r}')
    if timeout < 0: raise TypeError(f'timeout must be nonnegative; received: {timeout!r}')
    self.timeout = timeout
    self.msg = msg
    self.on_signal = on_signal
    self.timed_out = False


  def __enter__(self) -> None:
    if not self.timeout: return
    prev = signal.getsignal(signal.SIGALRM)
    if prev is not signal.SIG_DFL:
      raise Exception(f'task.run encountered previously installed signal handler: {prev}')

    def alarm_handler(signum, current_stack_frame) -> None:
      # since signal handlers carry reentrancy concerns; do not do any IO within the handler.
      self.timed_out = True
      if self.on_signal: self.on_signal() # callback must respect reentrancy limitations.

    signal.signal(signal.SIGALRM, alarm_handler) # set handler.
    signal.alarm(self.timeout) # set alarm.


  def __exit__(self, exc_type:OptTypeBaseExc, exc_value:OptBaseExc, traceback:OptTraceback) -> None:
    if not self.timeout: return
    signal.alarm(0) # disable alarm.
    signal.signal(signal.SIGALRM, signal.SIG_DFL) # uninstall handler.
    if exc_type: return # exception will be reraised.
    if self.timed_out: raise Timeout(self.msg.format(timeout=self.timeout))
