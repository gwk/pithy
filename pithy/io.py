# Dedicated to the public domain under CC0: https://creativecommons.org/publicdomain/zero/1.0/.

from os import close as os_close, O_NONBLOCK, O_RDONLY, open as os_open, read as os_read
from pprint import pprint
from shlex import quote as sh_quote
from string import Template as _Template
from sys import stderr, stdin, stdout
from typing import Any, Callable, cast, ContextManager, Iterable, Iterator, Sized, TextIO, TypeVar

from .desc import errD, outD, writeD
from .reprs import repr_ml
from .typing import OptBaseExc, OptTraceback, OptTypeBaseExc


errD = errD
outD = outD
writeD = writeD


_T = TypeVar('_T')

# basic printing.

def writeZ(file: TextIO, *items: Any, sep='', end='', flush=False) -> None:
  "Write `items` to file; default sep='', end=''."
  print(*items, sep=sep, end=end, file=file, flush=flush)

def writeS(file: TextIO, *items: Any, sep='', flush=False) -> None:
  "Write `items` to file; sep='', end=' '."
  print(*items, sep=sep, end=' ', file=file, flush=flush)

def writeSZ(file: TextIO, *items: Any, flush=False) -> None:
  "Write `items` to file; sep=' ', end=''."
  print(*items, sep=' ', end='', file=file, flush=flush)

def writeSS(file: TextIO, *items: Any, flush=False) -> None:
  "Write `items` to file; sep=' ', end=''."
  print(*items, sep=' ', end=' ', file=file, flush=flush)

def writeL(file: TextIO, *items: Any, sep='', flush=False) -> None:
  "Write `items` to file; sep='', end='\\n'."
  print(*items, sep=sep, end='\n', file=file, flush=flush)

def writeN(file: TextIO, *items: Any, sep='', flush=False) -> None:
  "Write `items` to file; sep='', end=('' if items and items[-1].endswith('\\n') else '\\n')."
  end = ('' if items and items[-1].endswith('\n') else '\n')
  print(*items, sep=sep, end=end, file=file, flush=flush)

def writeSL(file: TextIO, *items: Any, flush=False) -> None:
  "Write `items` to file; sep=' ', end='\\n'."
  print(*items, sep=' ', end='\n', file=file, flush=flush)

def writeSN(file: TextIO, *items: Any, flush=False) -> None:
  "Write `items` to file; sep='', end=('' if items and items[-1].endswith('\\n') else '\\n')."
  end = ('' if items and items[-1].endswith('\n') else '\n')
  print(*items, sep=' ', end=end, file=file, flush=flush)

def writeLL(file: TextIO, *items: Any, flush=False) -> None:
  "Write `items` to file; sep='\\n', end='\\n'."
  print(*items, sep='\n', end='\n', file=file, flush=flush)

def writeLSSL(file: TextIO, *items: Any, flush=False) -> None:
  "Write `items` to file; sep='\\n  ', end='\\n'."
  print(*items, sep='\n  ', end='\n', file=file, flush=flush)


# templated format printing.

def writeTF(file:TextIO, template_fmt:str, *items:Any, flush=False, **keyed_items:Any) -> None:
  """
  Expand the format string with keyed_items, then format the string; end=''.
  Useful for constructing dynamic format strings.
  """
  fmt = _Template(template_fmt).substitute(**keyed_items)
  print(fmt.format(*items, **keyed_items, end='', file=file, flush=flush))


def writeTFL(file:TextIO, template_fmt:str, *items:Any, flush=False, **keyed_items:Any) -> None:
  """
  Expand the format string template with keyed_items, then format the string; end='\\n'
  Useful for constructing dynamic format strings.
  """
  fmt = _Template(template_fmt).substitute(**keyed_items)
  print(fmt.format(*items, **keyed_items, file=file, flush=flush))


# Pretty printing.

def writeP(file:TextIO, *labels_and_obj: Any, indent=2, **opts:Any) -> None:
  'Write labels and pretty-print object to file.'
  labels = labels_and_obj[:-1]
  obj = labels_and_obj[-1]
  if labels: print(*labels, end=': ', file=file)
  pprint(obj, stream=file, indent=indent, **opts)


def writeM(file:TextIO, *labels_and_obj:Any, at_line_start:bool|None=None, color:bool|None=None, **opts:Any) -> None:
  'Write labels and multiline repr of object to file.'
  labels = labels_and_obj[:-1]
  obj = labels_and_obj[-1]
  if labels: print(*labels, end=': ', file=file)
  if at_line_start is None: at_line_start = not labels
  if color is None: color = file.isatty()
  print(repr_ml(obj, at_line_start=at_line_start, color=color, **opts), file=file)


# std out.

def outZ(*items: Any, sep='', end='', flush=False) -> None:
  "Write `items` to std out; sep='', end=''."
  print(*items, sep=sep, end=end, flush=flush)

def outS(*items: Any, sep='', flush=False) -> None:
  "Write `items` to std out; sep='', end=' '."
  print(*items, sep=sep, end=' ', flush=flush)

def outSZ(*items: Any, flush=False) -> None:
  "Write `items` to std out; sep=' ', end=''."
  print(*items, sep=' ', end='', flush=flush)

def outSS(*items: Any, flush=False) -> None:
  "Write `items` to std out; sep=' ', end=' '."
  print(*items, end=' ', flush=flush)

def outL(*items: Any, sep='', flush=False) -> None:
  "Write `items` to std out; sep='', end='\\n'."
  print(*items, sep=sep, flush=flush)

def outN(*items: Any, sep='', flush=False) -> None:
  "Write `items` to std out; sep='', end=('' if items and items[-1].endswith('\\n') else '\\n')."
  end = ('' if items and items[-1].endswith('\n') else '\n')
  print(*items, sep=sep, end=end, flush=flush)

def outSL(*items: Any, flush=False) -> None:
  "Write `items` to std out; sep=' ', end='\\n'."
  print(*items, flush=flush)

def outSN(*items: Any, flush=False) -> None:
  "Write `items` to std out; sep='', end=('' if items and items[-1].endswith('\\n') else '\\n')."
  end = ('' if items and items[-1].endswith('\n') else '\n')
  print(*items, sep=' ', end=end, flush=flush)

def outLL(*items: Any, flush=False) -> None:
  "Write `items` to std out; sep='\\n', end='\\n'."
  print(*items, sep='\n', flush=flush)

def outLSSL(*items: Any, flush=False) -> None:
  "Write `items` to std out; sep='\\n  ', end='\\n'."
  print(*items, sep='\n  ', flush=flush)

def outR(*items: Any, sep='', is_tty:bool|None=None, flush=False) -> None:
  '''
  Write `items` to std out. sep=''.
  If `is_tty`, end=ERASE_LINE_F+'\\r'; otherwise, end='\\n'.
  If `is_tty` is None, it defaults to stdout.isatty().
  '''
  if is_tty is None: is_tty = stdout.isatty()
  print(*items, sep=sep, end=('\x1b[0K\r' if is_tty else '\n'), flush=flush)

def outSR(*items: Any, is_tty:bool|None=None, flush=False) -> None:
  '''
  Write `items` to std out; sep=' '.
  If `is_tty`, end=ERASE_LINE_F+'\\r'; otherwise, end='\\n'.
  If `is_tty` is None, it defaults to stdout.isatty().
  '''
  if is_tty is None: is_tty = stdout.isatty()
  print(*items, sep=' ', end=('\x1b[0K\r' if is_tty else '\n'), flush=flush)

def outP(*labels_and_obj:Any, **opts: Any) -> None:
  'Pretty print to std out.'
  writeP(stdout, *labels_and_obj, **opts)

def outM(*labels_and_obj:Any, **opts: Any) -> None:
  'Multiline repr to std out.'
  writeM(stdout, *labels_and_obj, **opts)


# std err.

def errZ(*items: Any, sep='', end='', flush=False) -> None:
  "Write items to std err; default sep='', end=''."
  print(*items, sep=sep, end=end, file=stderr, flush=flush)

def errS(*items: Any, sep='', flush=False) -> None:
  "Write items to std err; sep='', end=' '."
  print(*items, sep=sep, end=' ', file=stderr, flush=flush)

def errSZ(*items: Any, flush=False) -> None:
  "Write items to std err; sep=' ', end=''."
  print(*items, sep=' ', end='', file=stderr, flush=flush)

def errSS(*items: Any, flush=False) -> None:
  "Write items to std err; sep=' ', end=''."
  print(*items, sep=' ', end=' ', file=stderr, flush=flush)

def errL(*items: Any, sep='', flush=False) -> None:
  "Write items to std err; sep='', end='\\n'."
  print(*items, sep=sep, end='\n', file=stderr, flush=flush)

def errN(*items: Any, sep='', flush=False) -> None:
  "Write `items` to std err; sep='', end=('' if items and items[-1].endswith('\\n') else '\\n')."
  end = ('' if items and items[-1].endswith('\n') else '\n')
  print(*items, sep=sep, end=end, file=stderr, flush=flush)

def errSL(*items: Any, flush=False) -> None:
  "Write items to std err; sep=' ', end='\\n'."
  print(*items, sep=' ', end='\n', file=stderr, flush=flush)

def errSN(*items: Any, flush=False) -> None:
  "Write `items` to std err; sep='', end=('' if items and items[-1].endswith('\\n') else '\\n')."
  end = ('' if items and items[-1].endswith('\n') else '\n')
  print(*items, sep=' ', end=end, file=stderr, flush=flush)

def errLL(*items: Any, flush=False) -> None:
  "Write items to std err; sep='\\n', end='\\n'."
  print(*items, sep='\n', end='\n', file=stderr, flush=flush)

def errLSSL(*items: Any, flush=False) -> None:
  "Write items to std err; sep='\\n  ', end='\\n'."
  print(*items, sep='\n  ', end='\n', file=stderr, flush=flush)

def errP(*labels_and_obj:Any, **opts) -> None:
  'Pretty print to std err.'
  writeP(stderr, *labels_and_obj, **opts)

def errM(*labels_and_obj:Any, **opts) -> None:
  'Multiline repr to std err.'
  writeM(stderr, *labels_and_obj, **opts)


def err_progress(iterable: Iterable[_T], label='progress', *, suffix='', final_suffix='', every:int=1, frequency:float=0.1,
 limit=0) -> Iterator[_T]:
  '''
  For interactive terminals, return a generator that yields the elements of `iterable`
  and displays a progress indicator on std err.
  If `frequency` is > 0, print a message every `frequency` seconds.
  Otherwise, print a message every `every` elements.
  If either `frequency` or `every` are 0, or stderr is not a TTY, return `iterable` unchanged.
  '''
  if not (every >= 0): raise ValueError(f'every must be >= 0; received {every!r}.')
  if not (frequency >= 0): raise ValueError(f'frequency must be >= 0; received {frequency!r}.')

  if every == 0 or frequency == 0 or not stderr.isatty():
    return iter(iterable)

  ERASE_LINE = '\x1b[2K'

  if label is None:
    label = str(iterable)
  pre = f'{ERASE_LINE}\r◊ {label}: '
  post = (suffix and ' ' + suffix) + '…'
  final = f' {final_suffix}.' if final_suffix else '.'

  total = ''
  width = 0
  try: l = len(cast(Sized, iterable))
  except TypeError: pass
  else:
    ls = f'{l:,}'
    width = len(ls)
    total = '/' + ls

  if frequency > 0:
    from time import monotonic as timer  # on macOS, monotomic and perf_counter both perform slightly better than `time`.

    # On macOS, calling the timer on every iteration induces ~1.3x overhead for a `x += 1` loop,
    # compared to the same loop with a very large integer frequency.
    # Since most loop bodies are more expensive, this seems broadly acceptable.
    # However, if the user wants to reduce this overhead, they can specify both `every` and `frequency`.
    # In that case, we check only check the timer every `every` iterations.

    if every == 1:
      def err_progress_gen_frequency() -> Iterator[_T]:
        prev_t = -1.0
        completed_count = 0
        for el in iterable:
          if limit and completed_count == limit: break
          t = timer()
          d = t - prev_t
          if d >= frequency:
            print(f'{pre}{completed_count:{width},}{total}{post}', end='', file=stderr, flush=True)
            prev_t = t
          yield el
          completed_count += 1
        print(f'{pre}{completed_count:{width},}{total}{final}', file=stderr)
      return err_progress_gen_frequency()

    else:
      def err_progress_gen_frequency_every() -> Iterator[_T]:
        prev_t = -1.0
        completed_count = 0
        for el in iterable:
          if limit and completed_count == limit: break
          if completed_count % every == 0:
            t = timer()
            d = t - prev_t
            if d >= frequency:
              print(f'{pre}{completed_count:{width},}{total}{post}', end='', file=stderr, flush=True)
              prev_t = t
          yield el
          completed_count += 1
        print(f'{pre}{completed_count:{width},}{total}{final}', file=stderr)
      return err_progress_gen_frequency_every()

  def err_progress_gen_every() -> Iterator[_T]:
    completed_count = 0
    for el in iterable:
      if limit and completed_count >= limit: break
      if completed_count % every == 0:
        print(f'{pre}{completed_count:{width},}{total}{post}', end='', file=stderr, flush=True)
      yield el
      completed_count += 1
    print(f'{pre}{completed_count:{width},}{total}{final}', file=stderr)
  return err_progress_gen_every()


# convenience read/write.


def read_from_path(path: str, default: str|None=None) -> str:
  'Read all text from file at `path`.'
  try:
    with open(path) as f:
      return f.read()
  except (FileNotFoundError, IsADirectoryError):
    if default is None: raise
    return default


def read_line_from_path(path: str, line_index=0, keep_end=False, default: str|None=None) -> str:
  'Read a single line of text from file at `path`.'
  try:
    with open(path) as f:
      for i, line in enumerate(f):
        if i == line_index:
          return line if keep_end else line.rstrip('\n')
      if default is None: raise IndexError(line_index)
      return default
  except (FileNotFoundError, IsADirectoryError, PermissionError, UnicodeDecodeError):
    if default is None: raise
    return default


def write_to_path(path:str, text:str|bytes|bytearray) -> None:
  'Writes `string` to file at `path`.'
  if isinstance(text, str):
    with open(path, 'w') as f: f.write(text)
  else:
    with open(path, 'wb') as bf: bf.write(text)


# Opener utility.

def mk_opener(flags:int, mode=0o777, dir_fd:int|None=None) -> Callable[[str, int], int]:
  def _opener(path:str, _flags:int, mode=mode, dir_fd=dir_fd) -> int: return os_open(path,_flags&flags)
  return _opener

nonblock_opener = mk_opener(O_NONBLOCK)


# Nonblocking tools.

class AsyncLineReader(ContextManager):
  '''
  A file-like object for reading asynchronously from a file descriptor.
  '''

  def __init__(self, path:str):
    self.fd = os_open(path, O_RDONLY|O_NONBLOCK) # TODO: accept raw descriptor.
    self.buffer = bytearray()

  def __del__(self) -> None:
    self.close()

  def __exit__(self, exc_type:OptTypeBaseExc, exc_value:OptBaseExc, traceback:OptTraceback) -> None:
    self.close()

  def close(self) -> None:
    if self.fd >= 0:
      os_close(self.fd)
      self.fd = -1

  def readline(self) -> str:
    '''
    Attempt to return a complete line from the input stream.
    If there is not enough data available, return ''.
    '''
    line:bytes|bytearray

    # The buffer might already contain a complete line.
    buffer_line_end = self.buffer.find(0x0a) + 1 # 0 if no newline found.
    if buffer_line_end: # Buffer already contains a complete line.
      line = self.buffer[:buffer_line_end]
      del self.buffer[buffer_line_end:]
      return line.decode()

    # Read from the file descriptor until it returns nothing or we have a complete line.
    while True:
      try: data = os_read(self.fd, 4096)
      except BlockingIOError: return ''
      if not data: return ''
      line_end = data.find(0x0a) + 1 # 0 if no newline found.
      if not line_end: # No newline.
        self.buffer.extend(data)
        continue
      if self.buffer: # Have previous data.
        line_end += len(self.buffer)
        self.buffer.extend(data)
        line = self.buffer[:line_end]
        del self.buffer[:line_end]
      else: # No previous data.
        line = data[:line_end]
        self.buffer.extend(data[line_end:])
      return line.decode()


# misc.

def clip_newlines(iterable:Iterable[str]) -> Iterable[str]:
  for line in iterable:
    yield line.rstrip('\n')


def confirm(question:str) -> bool:
  '''
  Prompt the user to confirm a question with a "y" response.
  "y" returns True; any other response returns False.
  The question is printed to stdout, followed by a question mark and prompt.
  '''
  from .term import CBreakMode
  print(f'{question}? press "y" to confirm: ', end='', flush=True)
  with CBreakMode(): response = stdin.read(1)
  print(response)
  return (response == 'y')


def confirm_or_exit(question:str) -> None:
  '''
  Prompt the user to confirm a question with a "y" response.
  "y" returns; any other response exits with status 1.
  The question is printed to stdout, followed by a question mark and prompt.
  '''
  try:
    if not confirm(question): exit(1)
  except KeyboardInterrupt: exit(1)


def assert_eq(a: Any, b: Any):
  if a != b:
    raise AssertionError(f'not equal:\n  {a!r}\n  {b!r}')


def shell_cmd_str(cmd:Iterable[str]) -> str:
  return ' '.join(sh_quote(word) for word in cmd)


def tee_to_err(iterable:Iterable[_T], label:str='tee_to_err', transform:Callable[[_T],Any]|None=None) -> Iterator[_T]:
  for el in iterable:
    s = repr(el) if transform is None else str(transform(el))
    errL(label, ': ', s)
    yield el
