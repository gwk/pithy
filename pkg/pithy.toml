
[project]

name = "pithy"
description = "Pithy is a collection of utility libraries for Python 3."
requires-python = ">=3.14"

dependencies = [
  "brotli >= 1.1.0",
  "mypy >= 1.18.2",
  "mypy-extensions >= 1.1.0",
  "cryptography-46.0.2",
  "starlette-0.48.0",
  "types-PyYAML >= 6.0.12.20250915",
  "lxml >= 6.0.2",
  "blake3 >= 1.0.7",
  "types-lxml >= 2025.8.25",
  "msgpack >= 1.1.2",
  "msgpack-types >= 0.5.0",
  "docutils >= 0.22.0",
]

keywords = [
  "ansi",
  "apfs",
  "buffer",
  "chart",
  "clone",
  "console",
  "crawl",
  "csv",
  "diff",
  "format",
  "graph",
  "graphviz",
  "html",
  "iterable",
  "json",
  "lex",
  "lexer",
  "lexing",
  "msgpack",
  "parse",
  "parser",
  "parsing",
  "scrape",
  "scraping",
  "sqlite",
  "svg",
  "syntax",
  "test",
  "testing",
  "term",
  "terminal",
  "token",
  "tokenization",
  "tokenizer",
  "types",
  "typing",
  "unicode",
  "unit",
  "utilities",
  "xml",
  "zst",
  "zstd",
  "zstandard",
]

classifiers = [
  "Environment :: Console",
  "Intended Audience :: Developers",
  "Intended Audience :: Education",
  "Intended Audience :: Information Technology",
  "Intended Audience :: Science/Research",
  "License :: CC0 1.0 Universal (CC0 1.0) Public Domain Dedication",
  "Programming Language :: Python :: 3",
  "Topic :: Documentation",
  "Topic :: Education",
  "Topic :: Internet",
  "Topic :: Multimedia",
  "Topic :: Office/Business",
  "Topic :: Scientific/Engineering",
  "Topic :: Software Development",
  "Topic :: Software Development :: Build Tools",
  "Topic :: Software Development :: Code Generators",
  "Topic :: Software Development :: Documentation",
  "Topic :: Software Development :: Libraries",
  "Topic :: Software Development :: Pre-processors",
  "Topic :: Software Development :: Testing",
  "Topic :: Software Development :: User Interfaces",
  "Topic :: Software Development :: Version Control",
  "Topic :: Software Development :: Version Control :: Git",
  "Topic :: System",
  "Topic :: Text Processing",
  "Topic :: Text Processing :: Filters",
  "Topic :: Text Processing :: General",
  "Topic :: Text Processing :: Markup",
  "Topic :: Text Processing :: Markup :: HTML",
  "Topic :: Utilities",
  "Typing :: Typed",
]
