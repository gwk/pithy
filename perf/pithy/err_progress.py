from sys import argv
from pithy.io import err_progress

n = int(argv[1])
every = int(argv[2])
frequency = float(argv[3])

s = 0
for i in err_progress(range(n), every=every, frequency=frequency):
    s += i
print(s)
