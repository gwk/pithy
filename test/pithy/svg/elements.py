#!/usr/bin/env python3

from pithy.io import outZ
from pithy.svg import *


w = 512
h = 512
svg = Svg(w=w, h=h)
svg.grid(size=(w, h), stroke='#E0E0E0', fill='white', stroke_width=0.5)

y = 0

g = svg.g(transform=translate(y=y))
g.rect((0, 0), (64, 32), r=8, title='rect0: radius=8')
g.rect(x=80, y=0, w=64, h=32, r=(16, 8), stroke='gray', fill='white', title='rect1: radius=(16, 8)')
y += 48

g = svg.g(transform=translate(y=y))
g.circle((16, 16), 16, title='circle0')
g.circle(x=64, y=16, r=16, stroke='gray', fill='white', title='circle1')
y += 48

g = svg.g(transform=translate(y=y))
g.line((0, 0), (32, 32), stroke='black', title='line0')
g.line(x1=32, y1=0, x2=64, y2=32, stroke='black', title='line1')
y += 48

g = svg.g(transform=translate(y=y))
g.path([('M', 0, 0), ('L', 32, 32), ('l', 32, -32)], stroke='black', fill=None, title='path0')
y += 48

g = svg.g(transform=translate(y=y))
g.polyline([(0,0), (32,32), (64,0)], stroke='black', fill=None, title='polyline0')
y += 48

svg.append(Text('TEXT', x=0, y=y))
y += 16

# 4x4 checkerboard of black and white pixels.
img_base64 = 'data:image/png;base64,R0lGODdhBAAEAIAAAAAAAP///yH5BAQAAAAALAAAAAAEAAQAAAIFRHxnuAUAOw=='

svg.image((0, y), (16,16), href=img_base64, title='image0')
svg.image(x=32, y=y, w=32, h=32, href=img_base64, title='image1')
y += 48

g = svg.g(transform=translate(y=y))
g1 = g.symbol(id='sym', vw=32, vh=32)
g1.rect(pos=(8, 8), size=(16, 16), stroke='black', fill=None)
g1.rect(pos=(0, 0), size=(32, 32), stroke='black', fill=None)
g.use('sym', (0,0), size=32)
g.use('sym', (48,0), size=16)
y += 48

g = svg.g(transform=translate(x=16, y=y))
marker = g.marker(id='dot', pos=(4,4), size=8, vw=8, vh=8)
marker.circle((4, 4), r=4)
g.polyline([(0,0), (16,0), (32,0)], stroke='black', marker_start="url(#dot)", marker_mid="url(#dot)", marker_end="url(#dot)",
  title='Polyline with marker dots')
y += 32

outZ(*svg.render())
