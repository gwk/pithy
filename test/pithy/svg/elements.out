<svg xmlns='http://www.w3.org/2000/svg' width='512' height='512'>
<g class='grid' stroke='#E0E0E0' fill='white' stroke-width='0.5'>
<line x1='16' y1='0.0' x2='16' y2='512'/>
<line x1='32' y1='0.0' x2='32' y2='512'/>
<line x1='48' y1='0.0' x2='48' y2='512'/>
<line x1='64' y1='0.0' x2='64' y2='512'/>
<line x1='80' y1='0.0' x2='80' y2='512'/>
<line x1='96' y1='0.0' x2='96' y2='512'/>
<line x1='112' y1='0.0' x2='112' y2='512'/>
<line x1='128' y1='0.0' x2='128' y2='512'/>
<line x1='144' y1='0.0' x2='144' y2='512'/>
<line x1='160' y1='0.0' x2='160' y2='512'/>
<line x1='176' y1='0.0' x2='176' y2='512'/>
<line x1='192' y1='0.0' x2='192' y2='512'/>
<line x1='208' y1='0.0' x2='208' y2='512'/>
<line x1='224' y1='0.0' x2='224' y2='512'/>
<line x1='240' y1='0.0' x2='240' y2='512'/>
<line x1='256' y1='0.0' x2='256' y2='512'/>
<line x1='272' y1='0.0' x2='272' y2='512'/>
<line x1='288' y1='0.0' x2='288' y2='512'/>
<line x1='304' y1='0.0' x2='304' y2='512'/>
<line x1='320' y1='0.0' x2='320' y2='512'/>
<line x1='336' y1='0.0' x2='336' y2='512'/>
<line x1='352' y1='0.0' x2='352' y2='512'/>
<line x1='368' y1='0.0' x2='368' y2='512'/>
<line x1='384' y1='0.0' x2='384' y2='512'/>
<line x1='400' y1='0.0' x2='400' y2='512'/>
<line x1='416' y1='0.0' x2='416' y2='512'/>
<line x1='432' y1='0.0' x2='432' y2='512'/>
<line x1='448' y1='0.0' x2='448' y2='512'/>
<line x1='464' y1='0.0' x2='464' y2='512'/>
<line x1='480' y1='0.0' x2='480' y2='512'/>
<line x1='496' y1='0.0' x2='496' y2='512'/>
<line x1='0.0' y1='16' x2='512' y2='16'/>
<line x1='0.0' y1='32' x2='512' y2='32'/>
<line x1='0.0' y1='48' x2='512' y2='48'/>
<line x1='0.0' y1='64' x2='512' y2='64'/>
<line x1='0.0' y1='80' x2='512' y2='80'/>
<line x1='0.0' y1='96' x2='512' y2='96'/>
<line x1='0.0' y1='112' x2='512' y2='112'/>
<line x1='0.0' y1='128' x2='512' y2='128'/>
<line x1='0.0' y1='144' x2='512' y2='144'/>
<line x1='0.0' y1='160' x2='512' y2='160'/>
<line x1='0.0' y1='176' x2='512' y2='176'/>
<line x1='0.0' y1='192' x2='512' y2='192'/>
<line x1='0.0' y1='208' x2='512' y2='208'/>
<line x1='0.0' y1='224' x2='512' y2='224'/>
<line x1='0.0' y1='240' x2='512' y2='240'/>
<line x1='0.0' y1='256' x2='512' y2='256'/>
<line x1='0.0' y1='272' x2='512' y2='272'/>
<line x1='0.0' y1='288' x2='512' y2='288'/>
<line x1='0.0' y1='304' x2='512' y2='304'/>
<line x1='0.0' y1='320' x2='512' y2='320'/>
<line x1='0.0' y1='336' x2='512' y2='336'/>
<line x1='0.0' y1='352' x2='512' y2='352'/>
<line x1='0.0' y1='368' x2='512' y2='368'/>
<line x1='0.0' y1='384' x2='512' y2='384'/>
<line x1='0.0' y1='400' x2='512' y2='400'/>
<line x1='0.0' y1='416' x2='512' y2='416'/>
<line x1='0.0' y1='432' x2='512' y2='432'/>
<line x1='0.0' y1='448' x2='512' y2='448'/>
<line x1='0.0' y1='464' x2='512' y2='464'/>
<line x1='0.0' y1='480' x2='512' y2='480'/>
<line x1='0.0' y1='496' x2='512' y2='496'/>
<rect class='grid-border' fill='none' x='0.0' y='0.0' width='512' height='512'/>
</g>
<g transform='translate(0,0)'>
<rect x='0' y='0' width='64' height='32' rx='8' ry='8'><title>rect0: radius=8</title></rect>
<rect stroke='gray' fill='white' x='80' y='0' width='64' height='32' rx='16' ry='8'><title>rect1: radius=(16, 8)</title></rect>
</g>
<g transform='translate(0,48)'>
<circle cx='16' cy='16' r='16'><title>circle0</title></circle>
<circle stroke='gray' fill='white' cx='64' cy='16' r='16'><title>circle1</title></circle>
</g>
<g transform='translate(0,96)'>
<line stroke='black' x1='0' y1='0' x2='32' y2='32'><title>line0</title></line>
<line stroke='black' x1='32' y1='0' x2='64' y2='32'><title>line1</title></line>
</g>
<g transform='translate(0,144)'><path stroke='black' fill='none' d='M0,0 L32,32 l32,-32'><title>path0</title></path></g>
<g transform='translate(0,192)'><polyline stroke='black' fill='none' points='0,0 32,32 64,0'><title>polyline0</title></polyline></g>
<text x='0' y='240'>TEXT</text>
<image href='data:image/png;base64,R0lGODdhBAAEAIAAAAAAAP///yH5BAQAAAAALAAAAAAEAAQAAAIFRHxnuAUAOw==' x='0' y='256' width='16' height='16'><title>image0</title></image>
<image href='data:image/png;base64,R0lGODdhBAAEAIAAAAAAAP///yH5BAQAAAAALAAAAAAEAAQAAAIFRHxnuAUAOw==' x='32' y='256' width='32' height='32'><title>image1</title></image>
<g transform='translate(0,304)'>
<symbol id='sym' viewBox='0 0 32 32'>
<rect stroke='black' fill='none' x='8' y='8' width='16' height='16'/>
<rect stroke='black' fill='none' x='0' y='0' width='32' height='32'/>
</symbol>
<use href='#sym' x='0' y='0' width='32' height='32'/>
<use href='#sym' x='48' y='0' width='16' height='16'/>
</g>
<g transform='translate(16,352)'>
<marker id='dot' refX='4' refY='4' markerWidth='8' markerHeight='8' viewBox='0 0 8 8' markerUnits='strokeWidth' orient='auto'><circle cx='4' cy='4' r='4'/></marker>
<polyline stroke='black' marker-start='url(#dot)' marker-mid='url(#dot)' marker-end='url(#dot)' points='0,0 16,0 32,0'><title>Polyline with marker dots</title></polyline>
</g>
</svg>
